include/sys/barrier.h

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright (c) 2022 Alexandre Mergnat <amergnat@baylibre.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * @brief Public interface for barriers.
+ */
+#ifndef ZEPHYR_INCLUDE_BARRIER_H_
+#define ZEPHYR_INCLUDE_BARRIER_H_
+
+#ifndef _ASMLANGUAGE
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#if defined(ARCH_HAS_MEMORY_BARRIER)
+
+#if defined(CONFIG_ARM)
+#include <arch/arm/aarch32/barrier.h>
+#elif defined(CONFIG_ARM64)
+#include <arch/arm64/barrier.h>
+#elif defined(CONFIG_RISCV)
+#include <arch/riscv/barrier.h>
+#else
+#error "Missing memory barrier definitions"
+#endif /* CONFIG_ARM */
+
+#elif defined(__GNUC__)
+
+/**
+ * @defgroup barrier_apis Barriers API
+ * @ingroup kernel_apis
+ * @{
+ */
+
+/**
+ * @brief z_full_mb - Memory Barrier
+ *
+ * A full system memory barrier. All memory operations before the mb() in the
+ * instruction stream will be committed before any operations after the mb()
+ * are committed.
+ */
+#define z_full_mb()	__atomic_thread_fence(__ATOMIC_ACQ_REL)
```

Comment:
I'm not convinced this is correct.  The acquire/release semantics are different that read/write barriers, at least in the way they're specified.  They force ordering for access from a single context, but allow racing contexts to see writes to the same values in arbitrary order.  That's different.  I think.  (Though it depends on exactly how ARM barriers are specified, and I'm not an expert).

IMHO it might be safer here to make all the barriers in this file SEQ_CST (sequentially consistent), which enforces a consistent order for all operations as seen by all contexts, at the cost of the inability to distinguish read/write barriers. 

Also, note that our atomics API is already specifying SEQ_CST, so mixing models is likely to lead to confusion.

Reply:
I agree ACQ_REL doesn't match with our full barrier definition.

AFAIU, according to [1] and [2], I suggest the following implementation to optimize and distinguish read/write:
'
#define z_full_mb()	__atomic_thread_fence(__ATOMIC_SEQ_CST)
#define z_read_mb()	__atomic_thread_fence(__ATOMIC_ACQUIRE)
#define z_write_mb()	__atomic_thread_fence(__ATOMIC_ACQ_REL)
'
These shouldn't allow data race.
I add more details about optimization on this thread: https://github.com/zephyrproject-rtos/zephyr/pull/42625#issuecomment-1068199825
If you think it still a bad suggestion after red the thread above, I'm ok to put __ATOMIC_SEQ_CST for read & write too.

[1] https://www.cplusplus.com/reference/atomic/atomic_thread_fence/
[2] https://en.cppreference.com/w/cpp/atomic/memory_order

Reply:
Yeah, I think I'd prefer we do this conservatively first and mark all barriers SEQ_CST.  Then later on someone with deep understanding of the C++ memory model can come along and "optimize" it by convincing us of the safety.